importer/src/path_canonicalizer.rs

```diff
@@ -0,0 +1,378 @@
+    ModuleStatement, Symbol, SymbolDefinition, SymbolPath,
+};
+
+pub fn canonicalize_paths<T: FieldElement>(
+    program: ASMProgram<T>,
+) -> Result<ASMProgram<T>, String> {
+    let paths = generate_path_map(&program)?;
+
+    let program = ASMProgram {
+        main: canonicalize_paths_in_module(AbsoluteSymbolPath::default(), program.main, &paths),
+    };
+
+    Ok(program)
+}
+
+// replace references to symbols with absolute paths. This removes the import statements. This always succeeds if the symbol table was generated correctly.
+fn canonicalize_paths_in_module<T: FieldElement>(
+    absolute_path: AbsoluteSymbolPath,
+    module: ASMModule<T>,
+    paths: &PathMap,
+) -> ASMModule<T> {
+    ASMModule {
+        statements: module
+            .statements
+            .into_iter()
+            .filter_map(|statement| match statement {
+                ModuleStatement::SymbolDefinition(SymbolDefinition { name, symbol }) => {
+                    match symbol {
+                        Symbol::Machine(m) => Some(Symbol::Machine(canonicalize_machine(
+                            absolute_path.clone(),
+                            m,
+                            paths,
+                        ))),
+                        Symbol::Import(_) => None,
+                        Symbol::Module(m) => match m {
+                            Module::External(_) => unreachable!(),
+                            Module::Local(m) => {
+                                Some(Symbol::Module(Module::Local(canonicalize_paths_in_module(
+                                    absolute_path.clone().join(name.clone()),
+                                    m,
+                                    paths,
+                                ))))
+                            }
+                        },
+                    }
+                    .map(|symbol| SymbolDefinition { name, symbol }.into())
+                }
+            })
+            .collect(),
+    }
+}
+
+fn canonicalize_machine<T: FieldElement>(
+    absolute_path: AbsoluteSymbolPath,
+    mut m: Machine<T>,
+    symbols: &PathMap,
+) -> Machine<T> {
+    for s in &mut m.statements {
+        if let MachineStatement::Submachine(_, path, _) = s {
+            *path = symbols
+                .get(&(absolute_path.clone(), std::mem::take(path)))
+                .cloned()
+                .unwrap()
+                .into();
+        }
+    }
+
+    m
+}
+
+/// Answers the question: When at absolute path `p`, refering to relative path `r`, what is the absolute path to the canonical symbol imported?
+pub type PathMap = BTreeMap<(AbsoluteSymbolPath, SymbolPath), AbsoluteSymbolPath>;
+
+#[derive(PartialEq, Debug)]
+pub struct State<'a, T> {
+    /// the root module of this program
+    root: &'a ASMModule<T>,
+    /// the paths to replace
+    pub paths: PathMap,
+}
+
+/// Checks a relative path in the context of an absolute path
```

Comment:
And especially, builds up the state. What does "check" mean here?

Reply:
So it's more like "resolves relative paths and stores the absolute paths in the state", right?

Reply:
changed doc